cluster-autoscaler/cloudprovider/openstack/openstack_nodegroup.go

```diff
@@ -0,0 +1,288 @@
+	if err != nil {
+		return fmt.Errorf("can not increase node count: %v", err)
+	}
+	if !updatePossible {
+		return fmt.Errorf("can not add nodes, cluster is in %s status", currentStatus)
+	}
+	klog.V(0).Infof("Increasing size by %d, %d->%d", delta, *ng.targetSize, *ng.targetSize+delta)
+	*ng.targetSize += delta
+
+	err = ng.openstackManager.UpdateNodeCount(ng.id, *ng.targetSize)
+	if err != nil {
+		return fmt.Errorf("could not increase cluster size: %v", err)
+	}
+
+	// Block until cluster has gone into update status and then completed
+	err = ng.WaitForClusterStatus(clusterStatusUpdateInProgress, waitForUpdateStatusTimeout)
+	if err != nil {
+		return fmt.Errorf("scale up error: %v", err)
+	}
+	err = ng.WaitForClusterStatus(clusterStatusUpdateComplete, waitForCompleteStatusTimout)
+	if err != nil {
+		return fmt.Errorf("scale up error: %v", err)
+	}
+	return nil
+}
+
+// DeleteNodes deletes a set of nodes chosen by the autoscaler.
+//
+// The process of deletion depends on the implementation of OpenstackManager,
+// but this function handles what should be common between all implementations:
+//   - simultaneous but separate calls from the autoscaler are batched together
+//   - does not allow scaling while the cluster is already in an UPDATE_IN_PROGRESS state
+//   - after scaling down, blocks until the cluster has reached UPDATE_COMPLETE
+func (ng *OpenstackNodeGroup) DeleteNodes(nodes []*apiv1.Node) error {
+
+	// Batch simultaneous deletes on individual nodes
+	ng.nodesToDeleteMutex.Lock()
+	// Check that these nodes would not make the batch delete more nodes than the minimum would allow
+	if *ng.targetSize-len(ng.nodesToDelete)-len(nodes) < ng.MinSize() {
+		ng.nodesToDeleteMutex.Unlock()
+		return fmt.Errorf("deleting nodes would take nodegroup below minimum size")
+	}
+	ng.nodesToDelete = append(ng.nodesToDelete, nodes...)
+	ng.nodesToDeleteMutex.Unlock()
+
+	ng.clusterUpdateMutex.Lock()
+	defer ng.clusterUpdateMutex.Unlock()
+
+	ng.nodesToDeleteMutex.Lock()
+	if len(ng.nodesToDelete) == 0 {
+		// Deletion was handled by another goroutine
+		ng.nodesToDeleteMutex.Unlock()
+		return nil
+	}
+	ng.nodesToDeleteMutex.Unlock()
+
+	// This goroutine has the clusterUpdateMutex, so will be the one
+	// to actually delete the nodes. While this goroutine waits, others
+	// will add their nodes to nodesToDelete and block at acquiring
+	// the clusterUpdateMutex lock. One they get it, the deletion will
+	// already be done and they will return above at the check
+	// for len(ng.nodesToDelete) == 0.
+	time.Sleep(deleteNodesBatchingDelay)
+
+	ng.nodesToDeleteMutex.Lock()
+	nodes = make([]*apiv1.Node, len(ng.nodesToDelete))
+	copy(nodes, ng.nodesToDelete)
+	ng.nodesToDelete = nil
+	ng.nodesToDeleteMutex.Unlock()
+
+	var nodeNames []string
+	for _, node := range nodes {
+		nodeNames = append(nodeNames, node.Name)
+	}
+	klog.V(1).Infof("Deleting nodes: %v", nodeNames)
+
+	updatePossible, currentStatus, err := ng.openstackManager.CanUpdate()
```

Comment:
Similarly to the IncreaseSize function, they stack/cluster should be able to handle multiple concurrent updates, as long as the intended result of the scaling action is stored and the subsequent action can act on that data.